core/vm/contracts.go

```diff
@@ -402,98 +416,194 @@ func modexpMultComplexity(x *big.Int) *big.Int {
 	return x
 }
 
+// berlinMultComplexity implements the multiplication complexity formula for Berlin.
+//
+// def mult_complexity(x):
+//
+//	ceiling(x/8)^2
+//
+// where is x is max(length_of_MODULUS, length_of_BASE)
+func berlinMultComplexity(x *big.Int) *big.Int {
+	x = new(big.Int).Add(x, big7)       // x + 7
+	x = new(big.Int).Rsh(x, 3)          // (x + 7) / 8
+	return new(big.Int).Mul(x, x)       // ((x + 7) / 8) ^ 2
+}
+
+// osakaMultComplexity implements the multiplication complexity formula for Osaka.
+//
+// For x <= 32: returns 16
+// For x > 32: returns 2 * ceiling(x/8)^2
+func osakaMultComplexity(x *big.Int) *big.Int {
+	if x.Cmp(big32) <= 0 {
+		return big.NewInt(16)
+	}
+	// For x > 32, return 2 * berlinMultComplexity(x)
+	berlinComplexity := berlinMultComplexity(x)
+	return new(big.Int).Lsh(berlinComplexity, 1) // 2 * berlinComplexity
+}
+
+// calculateIterationCount calculates the number of iterations for the modexp precompile.
+// This is the adjusted exponent length used in gas calculation.
+func calculateIterationCount(expLen uint64, expHead *big.Int, multiplier uint64) uint64 {
+	var iterationCount uint64
+	
+	if expLen <= 32 && expHead.Sign() == 0 {
+		iterationCount = 0
+	} else if expLen <= 32 {
+		// For small exponents, use MSB position - 1
+		if bitLen := expHead.BitLen(); bitLen > 0 {
+			iterationCount = uint64(bitLen - 1)
+		}
+	} else {
+		// For large exponents: (expLen - 32) * multiplier + MSB position - 1
+		iterationCount = (expLen - 32) * multiplier
+		if bitLen := expHead.BitLen(); bitLen > 0 {
+			iterationCount += uint64(bitLen - 1)
+		}
+	}
```

Comment:
This whole block is weird in two ways: 

- `iterationCount` is already zero upon declaration, so explicitly assigning zero is not needed
- the block with `BitLen` is duplicated

I would write it like this:

```go
if expLen > 32 && expHead.Sign() > 0 {
    iterationCount = (expLen - 32) * multiplier
}
if bitLen := expHead.BitLen(); bitLen > 0 {
    iterationCount += uint64(bitLen - 1)
}
```